### PR TITLE
Added raster zonal statistics section in Ch 04

### PR DESCRIPTION
**Pull request by michaeldorman**
Added section on zonal statistics

### CODE REVIEW

File: 04-spatial-operations.qmd [hunk 1]
```diff
@@ -494,12 +495,46 @@ Terrain processing, the calculation of topographic characteristics such as slope
 * `planform_curvature` 	Zevenbergen and Thorne (1987) doi: 10.1002/esp.3290120107
 * `profile_curvature` 	Zevenbergen and Thorne (1987) doi: 10.1002/esp.3290120107
 
+### Zonal operations
 
+Just like focal operations, zonal operations apply an aggregation function to multiple raster cells. However, a second raster, usually with categorical values, defines the zonal filters (or 'zones') in the case of zonal operations, as opposed to a predefined neighborhood window in the case of focal operation presented in the previous section. Consequently, raster cells defining the zonal filter do not necessarily have to be neighbors. Our grain size raster is a good example, as illustrated in the right panel of Figure 3.2: different grain sizes are spread irregularly throughout the raster. Finally, the result of a zonal operation is a summary table grouped by zone which is why this operation is also known as zonal statistics in the GIS world. This is in contrast to focal operations which return a raster object.
```
Comment (Robinlovelace) [inline 04-spatial-operations.qmd:500]: I think our style is sentence based breaks, so each sentence is on a new line, e.g.:

```
Just like focal operations, zonal operations apply an aggregation function to multiple raster cells.
However, a second raster, usually with categorical values, defines the zonal filters (or 'zones') in the case of zonal operations, as opposed to a predefined neighborhood window in the case of focal operation presented in the previous section.
Consequently, raster cells defining the zonal filter do not necessarily have to be neighbors.
Our grain size raster is a good example, as illustrated in the right panel of Figure 3.2: different grain sizes are spread irregularly throughout the raster.
Finally, the result of a zonal operation is a summary table grouped by zone which is why this operation is also known as zonal statistics in the GIS world.
This is in contrast to focal operations which return a raster object.
```

File: 04-spatial-operations.qmd [hunk 2]
```diff
@@ -494,12 +495,46 @@ Terrain processing, the calculation of topographic characteristics such as slope
...
+Our interntion is to calculate the average (or any other summary function, for that matter) of *elevation* in each zone defined by *grain* values. First, we can obtain the unique values defining the zones using `np.unique`: 
+
+```{python}
+np.unique(grain)
+```
+
+Now, we can use [dictionary conprehension](https://stackoverflow.com/questions/14507591/python-dictionary-comprehension) to "split" the `elev` array into separate one-dimensional arrays with values per `grain` group, with keys being the unique `grain` values:
+
+```{python}
+z = {i: elev[grain == i] for i in np.unique(grain)}
```
Comment (Robinlovelace) [inline 04-spatial-operations.qmd:524]: Nice syntax

**Review by Robinlovelace — APPROVED** [review]
Looks good and :+1: from me. I suggest changing the prose to sentence based breaks but we can do style/tidy stuff like that post merge.

File: 04-spatial-operations.qmd (hunk 1 above)
Comment (michaeldorman) [reply 04-spatial-operations.qmd:500]: Sure, thanks!

**Pull request merged**